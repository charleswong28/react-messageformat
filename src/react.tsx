import * as React from "react";
import { Token } from "messageformat-parser";
import { parse } from "./parse";
import { Values, evaluate } from "./eval";

export interface ContextValue {
  locale: string;
  compile: (id: string) => Token[];
}

export interface ProviderProps {
  locale: string;
  messageByID: { [key: string]: string | undefined };
  children?: React.ReactNode;
}

interface ProviderState {
  tokensByID: { [key: string]: Token[] | undefined };
}

export interface MessageOwnProps {
  id: string;
  values?: Values;
}

interface MessageContextProps {
  locale: string;
  compile: (id: string) => Token[];
}

type MessageProps = MessageOwnProps & MessageContextProps;

const defaultValue: ContextValue = {
  locale: "en",
  compile: () => [],
};
const { Provider: Provider_, Consumer } = React.createContext<ContextValue>(
  defaultValue
);

export class Provider extends React.Component<ProviderProps, ProviderState> {
  state: ProviderState = {
    tokensByID: {},
  };

  componentDidUpdate(prevProps: ProviderProps) {
    if (this.props.locale !== prevProps.locale) {
      this.setState({
        tokensByID: {},
      });
    }
  }

  compile = (id: string) => {
    const cachedTokens = this.state.tokensByID[id];
    if (cachedTokens != null) {
      return cachedTokens;
    }
    const source = this.props.messageByID[id];
    if (source == null) {
      throw new Error(`expected "${id}" to exist`);
    }
    const tokens = parse(source);
    this.setState(prevState => {
      return {
        tokensByID: {
          ...prevState.tokensByID,
          [id]: tokens,
        },
      };
    });
    return tokens;
  };

  render() {
    const { locale, children } = this.props;
    return (
      <Provider_
        value={{
          locale,
          compile: this.compile,
        }}
      >
        {children}
      </Provider_>
    );
  }
}

function Message(props: MessageProps) {
  const { id, values, locale, compile } = props;
  const tokens = compile(id);
  const result = evaluate(tokens, locale, values);
  const children = [];
  for (let i = 0; i < result.length; ++i) {
    const element = result[i];
    children.push(<React.Fragment key={i}>{element}</React.Fragment>);
  }
  return <>{children}</>;
}

export class FormattedMessage extends React.Component<MessageOwnProps> {
  renderMessage = (contextValue: ContextValue) => {
    const { id, values } = this.props;
    const { locale, compile } = contextValue;
    return (
      <Message id={id} values={values} locale={locale} compile={compile} />
    );
  };

  render() {
    return <Consumer>{this.renderMessage}</Consumer>;
  }
}
