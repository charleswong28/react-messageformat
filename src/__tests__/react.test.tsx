import * as React from "react";
import * as renderer from "react-test-renderer";
import {
  Provider,
  FormattedMessage,
  FormattedDate,
  Consumer,
  MessageOwnProps,
} from "../react";

const locale = "en";
const messageByID = {
  "plain.string": "Hello world",
  argument: "I meet {GUEST}",
  plural: "{N, plural, =4{four} one{ichi} other{#}}",
  "react.element": "{ANY} {NESTED} {REACT} {ELEMENT}",
  react: "This is a {A, react, href{{SCHEME}://{HOST}} children{link}}",
};

function Shortcut(props: MessageOwnProps) {
  return (
    <Provider locale={locale} messageByID={messageByID}>
      <FormattedMessage {...props} />
    </Provider>
  );
}

function Input(props: MessageOwnProps) {
  return (
    <Consumer>
      {({ renderToString }) => {
        return <input placeholder={renderToString(props.id, props.values)} />;
      }}
    </Consumer>
  );
}

class AnyComponent extends React.Component {
  render() {
    return "AnyComponent";
  }
}

test("plain.string", () => {
  const tree = renderer.create(<Shortcut id="plain.string" />).toJSON();
  expect(tree).toMatchSnapshot();
});

test("argument", () => {
  const tree = renderer
    .create(
      <Shortcut
        id="argument"
        values={{
          GUEST: "John",
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("plural", () => {
  const tree = renderer
    .create(
      <Shortcut
        id="plural"
        values={{
          N: 4,
        }}
      />
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("React Element as value", () => {
  const tree = renderer
    .create(
      <Provider locale={locale} messageByID={messageByID}>
        <FormattedMessage
          id="react.element"
          values={{
            ANY: <FormattedMessage id="plain.string" />,
            NESTED: <AnyComponent />,
            REACT: (
              <FormattedMessage
                id="argument"
                values={{
                  GUEST: "John",
                }}
              />
            ),
            ELEMENT: (
              <FormattedMessage
                id="plural"
                values={{
                  N: 1,
                }}
              />
            ),
          }}
        />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("Embedded React Element", () => {
  const tree = renderer
    .create(
      <Provider locale={locale} messageByID={messageByID}>
        <FormattedMessage
          id="react"
          components={{
            A: "a",
          }}
          values={{
            SCHEME: "https",
            HOST: "www.example.com",
          }}
        />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("imperative", () => {
  const tree = renderer
    .create(
      <Provider locale={locale} messageByID={messageByID}>
        <Input id="plain.string" />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});

test("FormattedDate", () => {
  const tree = renderer
    .create(
      <Provider locale={locale} messageByID={messageByID}>
        <FormattedDate value={new Date("2018-08-08")} month="long" />
      </Provider>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
