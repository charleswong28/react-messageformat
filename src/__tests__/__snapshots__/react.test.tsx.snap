// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`argument 1`] = `
Array [
  "I meet ",
  "John",
]
`;

exports[`plain.string 1`] = `"Hello world"`;

exports[`plural 1`] = `"four"`;

exports[`react.element 1`] = `
Array [
  "Hello world",
  " ",
  "AnyComponent",
  " ",
  "I meet ",
  "John",
  " ",
  "ichi",
]
`;
